packages/caliper-core/test/worker/rate-control/fixedFeedbackRate.js

```diff
@@ -108,41 +108,176 @@ describe('fixedFeedbackRate controller implementation', () => {
+
+            // Simulate conditions where there are no successful transactions
+            txnStats.stats.txCounters.totalSubmitted = 5;
+            txnStats.stats.txCounters.totalFinished = 2;
+            txnStats.stats.txCounters.totalSuccessful = 0;
+
+            // Set zeroSuccessfulCounter to 31
+            controller.zeroSuccessfulCounter = 31;
+
+            // Use the fake timers to control Date.now()
+            clock.tick(0); // Ensure clock starts at 0
+
+            await controller.applyRateControl();
+
+            sinon.assert.calledOnce(sleepStub);
+            // Expected sleep time should be capped at 30 * backOffTime = 3000ms
+            sinon.assert.calledWith(sleepStub, 30 * controller.backOffTime);
+        });
+
+        it('should sleep to adjust rate when expected time exceeds actual elapsed time', async () => {
+            txnStats.stats.txCounters.totalSubmitted = 44;
+            txnStats.stats.txCounters.totalFinished = 40;
+
+            // Simulate time progression
+            const elapsedTime = 1000; // e.g., 1000ms since start
+            clock.tick(elapsedTime);
+
+            await controller.applyRateControl();
+
+            sinon.assert.calledOnce(sleepStub);
+            // Calculate expected adjustment time
+            const expectedTime = controller.generalSleepTime * txnStats.getTotalSubmittedTx();
+            const actualElapsedTime = (Date.now() - controller.totalSleepTime) - txnStats.getRoundStartTime();
+            const adjustmentTime = expectedTime - actualElapsedTime;
+
+            sinon.assert.calledWith(sleepStub, adjustmentTime);
+        });
+
+        it('should sleep correctly based on the number of unfinished transactions', async () => {
+            // Configure the rate controller
+            testMessage.content.rateControl.opts = {
+                tps: 10000,
+                sleepTime: 100,
+                transactionLoad: 10
+            };
+
+            for (let i = 10; i > 0; --i) {
+                // Re-instantiate the controller for each iteration
+                controller = new FixedFeedbackRate.createRateController(testMessage, txnStats, 0);
+                const unfinishedPerWorker = controller.unfinishedPerWorker; // Should be 5
+
+                // Reset sleep stub
+                sleepStub.resetHistory();
+
+                // Ensure there are successful transactions
+                sinon.stub(txnStats, 'getTotalSuccessfulTx').returns(5);
+
+                // Stub getRoundStartTime
+                sinon.stub(txnStats, 'getRoundStartTime').returns(0);
+
+                // Stub getTotalSubmittedTx and getTotalFinishedTx
+                const unfinished = i * unfinishedPerWorker;
+                const totalFinished = 10;
+                const totalSubmitted = totalFinished + unfinished;
+
+                sinon.stub(txnStats, 'getTotalSubmittedTx').returns(totalSubmitted);
```

Comment:
again wonder why you stub here, but all others you set the txnStats values directly